Add validate command to cli

wilfredallyn opened the pull request:
Addresses https://github.com/fedimint/fedimint/issues/362

- Created function `validate_tokens` that verifies signatures by calling `Coin::verify` (does not check if nonce spent)
- Update cli with `Validate` command
- Updated readme with expected usage

Example usage:
```
$ mint-client-cli validate AQAAAAAAAABAQg8AAA...

All tokens have valid signatures
```

wilfredallyn commented on client/client-lib/src/lib.rs:
```diff
@@ -243,6 +243,21 @@ impl<T: AsRef<ClientConfig> + Clone> Client<T> {
+        let tbs_pks = &self.mint_client().context.config.tbs_pks;
+        if coins
+            .iter()
+            .map(|(amt, coin)| coin.coin.verify(*tbs_pks.tier(&amt).unwrap()))
```
I tried to use `?` instead of `unwrap` here but couldn't get it to compile. I was trying to do something similar to this https://github.com/fedimint/fedimint/blob/85e6602d6b3df79072bc999b1f966ef90d89d4f2/client/client-lib/src/mint/mod.rs#L262

wilfredallyn replied:
I added a variant to `ClientError` and replaced the `unwrap` with a `?`

wilfredallyn commented on client/client-lib/src/lib.rs:
```diff
@@ -243,6 +243,21 @@ impl<T: AsRef<ClientConfig> + Clone> Client<T> {
 
+    /// Validate tokens without claiming them. This function checks if signatures are valid
+    /// based on the federation public key. It does not check if the nonce is unspent.
+    pub async fn validate_tokens(&self, coins: Coins<SpendableCoin>) -> Result<()> {
```
I returned `Result<()>` but wasn't sure what type this should return

elsirion replied:
Returning `Result<()>` here is the right way. What you should do though is using a reference for the coins like so: `coins: &Coins<SpendableCoin>` so callers don't need to unnecessarily clone the struct.

wilfredallyn replied:
Thanks for the headsup. I revised the function definition to use a reference to coins

wilfredallyn commented on client/client-lib/src/lib.rs:
```diff
@@ -243,6 +243,21 @@ impl<T: AsRef<ClientConfig> + Clone> Client<T> {
         Ok(OutPoint { txid, out_idx: 0 })
     }
 
+    /// Validate tokens without claiming them. This function checks if signatures are valid
```
Throughout the comments and function naming, I used the term `tokens` instead of `coins` since I think it was mentioned on the devcall that the correct term is `tokens`. It may be confusing since other functions refer to `coins`. Let me know if I should rename anything.

wilfredallyn commented on client/cli/src/main.rs:
```diff
@@ -148,6 +154,9 @@ async fn main() {
             info!(%id, "Started reissuance, please fetch the result later");
         }
+        Command::Validate { coins } => {
+            client.validate_tokens(coins).await.expect("Invalid tokens");
```
I'm not sure if an `expect` is the right thing to do here

elsirion replied:
I think so, let's see if @dpc has a different view in #416

elsirion commented on docs/dev-running.md:
```diff
@@ -154,17 +162,21 @@ OPTIONS:
 SUBCOMMANDS:
+    connect-info      Config enabling client to establish websocket connection to federation
     fetch             Fetch (re-)issued coins and finalize issuance process
     help              Print this message or the help of the given subcommand(s)
     info              Display wallet info (holdings, tiers)
-    ln-pay            Pay a lightning invoice via a gateway
+    join-federation   Join a federation using it's ConnectInfo
     ln-invoice        Create a lightning invoice to receive payment via gateway
+    ln-pay            Pay a lightning invoice via a gateway
     peg-in            Issue tokens in exchange for a peg-in proof (not yet implemented, just
                           creates coins)
     peg-in-address    Generate a new peg-in address, funds sent to it can later be claimed
     peg-out           Withdraw funds from the federation
     reissue           Reissue tokens received from a third party to avoid double spends
     spend             Prepare coins to send to a third party as a payment
-    wait-block-height    Wait for the fed to reach a consensus block height
+    validate          Validate tokens without claiming them (only checks if signatures valid,
+                          does not check if nonce unspent)
+    wait-block-height Wait for the fed to reach a consensus block height
     wait-invoice      Wait for incoming invoice to be paid
```
Unrelated, just because we apparently forgot to update this: Is there a way to check that these are always up to date in CI?

wilfredallyn replied:
I added the description for the `validate` command and updated the other ones since I saw they were outdated. I don't know if CI can check this. Should I follow up with someone?

elsirion reviewed (changes requested):
Great PR! Just introduce a new error variant instead of unwrapping and this is good to go imo.

wilfredallyn commented:
I made the following changes based on feedback:
* added `ClientError` variant to replace unwrap
* revised function definition to use reference to coins
* fixed CI issues (fixed formatting and updated config since BorrowedClientContext removed in https://github.com/fedimint/fedimint/pull/448, )

elsirion reviewed (approved):
Thx!

elsirion commented:
I did some fixups of two nits I only noticed now (maybe I missed them the first time around)

justinmoon commented:
Nice PR @wilfredallyn!
